File: src/ovi_voice_assistant/device_connection.py
"""Unified device connection — bridges transport to the voice pipeline."""

from __future__ import annotations

import asyncio
import logging
import struct
from collections.abc import AsyncIterator, Awaitable, Callable
from typing import TYPE_CHECKING

from ovi_voice_assistant.agent.assistant_context import AssistantContext
from ovi_voice_assistant.codec import AudioCodec, create_codec
from ovi_voice_assistant.codec.lc3 import LC3_MUSIC_NBYTE
from ovi_voice_assistant.config import Settings
from ovi_voice_assistant.music import MusicGroup, MusicPlayer
from ovi_voice_assistant.pipeline_output import PipelineOutput
from ovi_voice_assistant.transport import AudioConfig, DeviceTransport, EventType
from ovi_voice_assistant.voice_assistant import VoiceAssistant

if TYPE_CHECKING:
    from ovi_voice_assistant.memory import Memory
    from ovi_voice_assistant.scheduler import Scheduler

# Callback signature for wake-word arbitration.
# (connection, score, wake_word) -> None
WakeCallback = Callable[["DeviceConnection", int, str], Awaitable[None]]

logger = logging.getLogger(__name__)


class _EncodingOutput(PipelineOutput):
    """Adapts transport + codec to PipelineOutput.

    The pipeline sends raw PCM; this encodes it with the codec
    and sends the encoded frames paced to real-time, with a small
    lead time so the device buffer can absorb network jitter.
    """

    LEAD_TIME = 0.3  # seconds — send this much audio ahead of real-time

    def __init__(self, transport: DeviceTransport, codec: AudioCodec) -> None:
        self._transport = transport
        self._codec = codec
        self._pcm_buf = b""
        self._frame_count = 0
        self._t0 = 0.0
        self._frame_duration = codec.frame_duration_ms / 1000.0

    async def send_event(self, event: EventType, payload: bytes = b"") -> None:
        # Send AUDIO_CONFIG before TTS_START so device configures decoder
        if event == EventType.TTS_START:
            config_payload = struct.pack(
                "<IHBB",
                self._codec.sample_rate,
                self._codec.encoded_frame_bytes,
                self._codec.codec_id,
                self._codec.channels,
            )
            await self._transport.send_event(EventType.AUDIO_CONFIG, config_payload)
        await self._transport.send_event(event, payload)

    async def send_audio(self, pcm: bytes) -> None:
        """Receive PCM from pipeline, encode with codec, pace to real-time."""
        loop = asyncio.get_running_loop()
        self._pcm_buf += pcm
        frame_bytes = self._codec.pcm_frame_bytes
        while len(self._pcm_buf) >= frame_bytes:
            frame = self._pcm_buf[:frame_bytes]
            self._pcm_buf = self._pcm_buf[frame_bytes:]
            encoded = self._codec.encode(frame)

            if self._frame_count == 0:
                self._t0 = loop.time()
            self._frame_count += 1

            await self._transport.send_audio(encoded)

            # Pace to real-time, but stay LEAD_TIME ahead so the device
            # buffer can absorb network jitter
            ahead = (self._frame_count * self._frame_duration) - (
                loop.time() - self._t0
            )
            if ahead > self.LEAD_TIME:
                await asyncio.sleep(ahead - self.LEAD_TIME)

    async def flush(self) -> None:
        """Flush any remaining PCM (pad to full frame)."""
        if self._pcm_buf:
            frame_bytes = self._codec.pcm_frame_bytes
            self._pcm_buf += b"\x00" * (frame_bytes - len(self._pcm_buf))
            encoded = self._codec.encode(self._pcm_buf)
            await self._transport.send_audio(encoded)
            self._pcm_buf = b""

    def reset(self) -> None:
        self._pcm_buf = b""
        self._frame_count = 0
        self._t0 = 0.0


class DeviceConnection:
    """Connects to a single Voice PE and routes audio through the pipeline."""

    def __init__(
        self,
        transport: DeviceTransport,
        spk_codec: AudioCodec,
        pipeline: VoiceAssistant,
        settings: Settings,
        on_wake: WakeCallback | None = None,
        name: str = "",
        music_group: MusicGroup | None = None,
    ) -> None:
        self._transport = transport
        self._spk_codec = spk_codec
        self._mic_codec: AudioCodec | None = (
            None  # created when device sends MIC_CONFIG
        )
        self._pipeline = pipeline
        self._settings = settings
        self._output = _EncodingOutput(transport, spk_codec)
        self._on_wake = on_wake
        self.name = name or repr(transport)

        self._audio_queue: asyncio.Queue[bytes | None] = asyncio.Queue()
        self._task: asyncio.Task | None = None
        self._continuing = False
        self._announcing = False
        self._context: AssistantContext | None = None
        self._music_output: _EncodingOutput | None = None
        self._music_group = music_group

    def set_scheduler(self, scheduler: Scheduler) -> None:
        """Attach the scheduler so automation tools are available."""
        if self._context:
            self._context.scheduler = scheduler

    def set_memory(self, memory: Memory) -> None:
        """Attach memory so remember/recall tools are available."""
        if self._context:
            self._context.memory = memory

    async def start(self) -> None:
        self._transport.set_event_callback(self._on_event)
        self._transport.set_audio_callback(self._on_audio)
        self._transport.set_disconnect_callback(self._on_disconnect)
        self._transport.set_connect_callback(self._on_reconnect)
        await self._transport.connect()
        await asyncio.sleep(0.5)
        await self._setup_device()

    async def stop(self) -> None:
        await self._cancel_task()
        await self._transport.disconnect()

    async def _setup_device(self) -> None:
        await self._transport.send_audio_config(
            AudioConfig(
                sample_rate=self._spk_codec.sample_rate,
                encoded_frame_bytes=self._spk_codec.encoded_frame_bytes,
                codec_type=self._spk_codec.codec_id,
                channels=self._spk_codec.channels,
            )
        )
        # Device responds with MIC_CONFIG after receiving AUDIO_CONFIG,
        # which configures self._mic_codec via _on_event.

        # Music gets its own 48 kHz codec so playback is CD-quality+
        # while TTS stays at whatever rate the voice codec uses.
        # AUDIO_CONFIG is re-sent before each TTS_START, so the device
        # reconfigures its decoder automatically.
        music_codec = create_codec(
            self._settings.transport.codec, 48000, channels=2, nbyte=LC3_MUSIC_NBYTE
        )
        self._music_output = _EncodingOutput(self._transport, music_codec)

        # Register this device's music output with the shared group
        if self._music_group is not None:
            self._music_group.add_device(self._music_output, self._transport)

        music_player = MusicPlayer(
            sample_rate=music_codec.sample_rate, channels=music_codec.channels
        )
        self._context = AssistantContext(
            announce=self.announce,
            music_player=music_player,
            music_group=self._music_group,
        )

    # -- Transport callbacks --

    async def _on_event(self, event: EventType, payload: bytes) -> None:
        if event == EventType.WAKE_WORD:
            await self._cancel_task()

            # Pause group music on ALL devices when any device wakes
            if self._music_group and self._music_group.is_active:
                await self._music_group.pause()

            if self._continuing:
                # Follow-up — already committed to this device, skip arbitration.
                logger.info("Follow-up session started")
                self._continuing = False
                self._audio_queue = asyncio.Queue()
                self._task = asyncio.create_task(self._run())
                return

            # Parse extended payload: [2B LE peak][2B LE ambient][wake_word UTF-8]
            # The normalised score (peak/ambient) cancels out mic gain
            # differences so the *closest* device wins, not the loudest mic.
            score = 0
            wake_word = ""
            if len(payload) >= 4:
                peak, ambient = struct.unpack_from("<HH", payload, 0)
                wake_word = payload[4:].decode("utf-8", errors="replace")
                score = peak * 1000 // max(ambient, 1)
                logger.debug(
                    "Wake payload: peak=%d ambient=%d → score=%d", peak, ambient, score
                )
            elif payload:
                wake_word = payload.decode("utf-8", errors="replace")

            # Prepare audio queue so mic frames arriving during
            # the arbitration window are buffered, not lost.
            self._audio_queue = asyncio.Queue()

            if self._on_wake is not None:
                await self._on_wake(self, score, wake_word)
            else:
                # No arbitration — start immediately.
                await self._pipeline.agent.reset_history()
                logger.info("Voice session started (wake_word=%r)", wake_word)
                self._task = asyncio.create_task(self._run())

        elif event == EventType.MIC_CONFIG:
            if len(payload) >= 7:
                rate = struct.unpack_from("<I", payload, 0)[0]
                nbyte = struct.unpack_from("<H", payload, 4)[0]
                codec_id = payload[6]
                codec_map = {0: "pcm", 1: "lc3", 2: "opus"}
                codec_name = codec_map.get(codec_id, "pcm")

                # Use the device's reported mic codec
                self._mic_codec = create_codec(codec_name, rate)
                logger.info(
                    "Device mic config: %s @ %dHz, %d bytes/frame",
                    codec_name,
                    rate,
                    nbyte,
                )

                # If the server prefers a different mic codec, request it
                preferred = self._settings.transport.codec
                if preferred != codec_name:
                    preferred_codec = create_codec(preferred, rate)
                    mic_config_payload = struct.pack(
                        "<IHB",
                        rate,
                        preferred_codec.encoded_frame_bytes,
                        preferred_codec.codec_id,
                    )
                    await self._transport.send_event(
                        EventType.MIC_CONFIG, mic_config_payload
                    )
                    self._mic_codec = preferred_codec
                    logger.info(
                        "Requested mic codec change: %s @ %dHz", preferred, rate
                    )

    async def _on_audio(self, data: bytes) -> None:
        """Decode mic audio and queue PCM for the pipeline."""
        if self._mic_codec is None:
            pcm = data
        else:
            pcm = self._mic_codec.decode(data)

        self._audio_queue.put_nowait(pcm)

    async def _on_disconnect(self) -> None:
        if self._task and not self._task.done():
            self._task.cancel()

    async def _on_reconnect(self) -> None:
        await asyncio.sleep(0.5)
        await self._setup_device()

    # -- Pipeline execution --

    async def _run(self) -> None:
        self._output.reset()
        self._continuing = await self._pipeline.run(
            self._output,
            self._mic_stream(),
            context=self._context,
        )
        await self._output.flush()

        # After the voice pipeline completes, resume music if it was playing
        if not self._continuing and self._context:
            if self._music_group and self._music_group.is_active:
                # Group music was paused for voice — resume on all devices
                await self._music_group.resume()
            elif self._context.music_player and self._context.music_player.is_active:
                await self._run_music()

    async def _run_music(self) -> None:
        """Stream music at 48 kHz via the dedicated music output."""
        player = self._context.music_player
        out = self._music_output
        out.reset()
        await out.send_event(EventType.TTS_START)
        try:
            await player.stream(out)
            await out.flush()
            await out.send_event(EventType.TTS_END)
        except asyncio.CancelledError:
            try:
                await out.flush()
                await out.send_event(EventType.TTS_END)
            except Exception:
                pass
            raise

    async def _run_announce(self, text: str) -> None:
        try:
            self._output.reset()
            await self._pipeline.announce(self._output, text)
            await self._output.flush()
        finally:
            self._announcing = False

    # -- Helpers --

    async def _mic_stream(self) -> AsyncIterator[bytes]:
        while True:
            chunk = await self._audio_queue.get()
            if chunk is None:
                break
            yield chunk

    async def _cancel_task(self) -> None:
        if self._task and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError, Exception:
                pass

    async def start_pipeline(self, wake_word: str) -> None:
        """Called by DeviceManager when this device wins arbitration."""
        await self._pipeline.agent.reset_history()
        logger.info(
            "Voice session started on %s (wake_word=%r, won arbitration)",
            self.name,
            wake_word,
        )
        self._task = asyncio.create_task(self._run())

    async def abort_wake(self) -> None:
        """Called by DeviceManager when this device loses arbitration."""
        await self._cancel_task()
        # Drain any buffered mic audio
        while not self._audio_queue.empty():
            try:
                self._audio_queue.get_nowait()
            except asyncio.QueueEmpty:
                break
        await self._transport.send_event(EventType.WAKE_ABORT)
        logger.info("Wake aborted on %s (another device won)", self.name)

    def announce(self, text: str) -> asyncio.Task:
        """Stream a TTS announcement directly to the device."""
        if self._task and not self._task.done():
            self._task.cancel()
        self._task = asyncio.ensure_future(self._run_announce(text))
        return self._task

File: internal/pipeline/device_connection.go
package pipeline

import (
	"context"
	"encoding/binary"
	"fmt"
	"github.com/bryfur/ovi-voice-assistant/internal/agent/scheduler"
	"github.com/bryfur/ovi-voice-assistant/internal/device"
	"github.com/bryfur/ovi-voice-assistant/internal/device/codec"
	"log/slog"
	"sync"
	"time"

	"github.com/bryfur/ovi-voice-assistant/internal/agent"
	"github.com/bryfur/ovi-voice-assistant/internal/config"
	"github.com/bryfur/ovi-voice-assistant/internal/music"
)

// WakeCallback is invoked for wake-word arbitration: (connection, score, wakeWord).
type WakeCallback func(conn *DeviceConnection, score int, wakeWord string)

// micQueueSize bounds buffered mic chunks (~80 s of 10 ms frames).
const micQueueSize = 8192

// Runner is the voice pipeline surface a connection needs.
type Runner interface {
	Run(ctx context.Context, output device.Output, mic <-chan []byte, actx *agent.Context) bool
	Announce(ctx context.Context, output device.Output, text string)
	ResetHistory()
}

// DeviceConnection connects to a single voice device and routes audio
// through the pipeline.
type DeviceConnection struct {
	Name string

	transport device.Transport
	spkCodec  codec.AudioCodec
	pipeline  Runner
	settings  *config.Settings
	output    *device.EncodingOutput
	onWake    WakeCallback

	mu          sync.Mutex
	micCodec    codec.AudioCodec // created when device sends MIC_CONFIG
	lastSpeaker string           // last speaker config announced to the device
	audioQueue  chan []byte
	taskCancel  context.CancelFunc
	taskDone    chan struct{}
	continuing  bool
	context     *agent.Context
	musicOutput *device.EncodingOutput
	musicGroup  *music.MusicGroup

	// SetupDelay is the pause after (re)connect before device setup.
	SetupDelay time.Duration
}

// DeviceOptions configure a DeviceConnection.
type DeviceOptions struct {
	OnWake     WakeCallback
	Name       string
	MusicGroup *music.MusicGroup
}

// NewDeviceConnection creates an unstarted connection.
func NewDeviceConnection(t device.Transport, spkCodec codec.AudioCodec, p Runner, settings *config.Settings, opts DeviceOptions) *DeviceConnection {
	name := opts.Name
	if name == "" {
		name = t.String()
	}
	c := &DeviceConnection{
		Name:       name,
		transport:  t,
		spkCodec:   spkCodec,
		pipeline:   p,
		settings:   settings,
		output:     device.NewEncodingOutput(t, spkCodec),
		onWake:     opts.OnWake,
		audioQueue: make(chan []byte, micQueueSize),
		musicGroup: opts.MusicGroup,
		SetupDelay: 500 * time.Millisecond,
	}
	c.output.OnAudioConfig = c.speakerConfig
	return c
}

// speakerConfig logs the speaker settings whenever they change (voice ↔ music).
func (c *DeviceConnection) speakerConfig(cd codec.AudioCodec) {
	desc := codec.Describe(cd)
	c.mu.Lock()
	changed := desc != c.lastSpeaker
	c.lastSpeaker = desc
	c.mu.Unlock()
	if changed {
		slog.Info("Speaker audio config", "device", c.Name, "codec", desc)
	}
}

// Context returns the agent context for this device (nil before setup).
func (c *DeviceConnection) Context() *agent.Context {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.context
}

// SetScheduler attaches the scheduler so automation tools are available.
func (c *DeviceConnection) SetScheduler(s *scheduler.Scheduler) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.context != nil {
		c.context.Scheduler = s
	}
}

// Start connects the transport and configures the device.
func (c *DeviceConnection) Start() error {
	c.transport.SetEventCallback(c.onEvent)
	c.transport.SetAudioCallback(c.onAudio)
	c.transport.SetDisconnectCallback(c.onDisconnect)
	c.transport.SetConnectCallback(c.onReconnect)
	if err := c.transport.Connect(); err != nil {
		return err
	}
	time.Sleep(c.SetupDelay)
	return c.setupDevice()
}

// Stop cancels any running task and disconnects.
func (c *DeviceConnection) Stop() error {
	c.cancelTask()
	return c.transport.Disconnect()
}

// setupDevice sends the speaker config and prepares the music path and
// agent context.
func (c *DeviceConnection) setupDevice() error {
	c.speakerConfig(c.spkCodec)
	if err := device.SendAudioConfig(c.transport, device.AudioConfig{
		SampleRate:        uint32(c.spkCodec.SampleRate()),
		EncodedFrameBytes: uint16(c.spkCodec.EncodedFrameBytes()),
		CodecType:         c.spkCodec.ID(),
		Channels:          uint8(c.spkCodec.Channels()),
	}); err != nil {
		return err
	}
	// Device responds with MIC_CONFIG after receiving AUDIO_CONFIG, which
	// configures micCodec via onEvent.

	c.mu.Lock()
	defer c.mu.Unlock()
	if c.context != nil {
		return nil // already set up (reconnect)
	}
	// Music gets its own 48 kHz codec so playback is CD-quality+ while TTS
	// stays at whatever rate the voice codec uses. AUDIO_CONFIG is re-sent
	// before each TTS_START, so the device reconfigures its decoder.
	musicCodec, err := codec.NewCodec(c.settings.Transport.Codec, 48000, 2, codec.LC3MusicNByte)
	if err != nil {
		return fmt.Errorf("music codec: %w", err)
	}
	c.musicOutput = device.NewEncodingOutput(c.transport, musicCodec)
	c.musicOutput.OnAudioConfig = c.speakerConfig
	slog.Info("Music audio config", "device", c.Name, "codec", codec.Describe(musicCodec))
	if c.musicGroup != nil {
		c.musicGroup.AddDevice(c.musicOutput, c.transport)
	}
	player := music.NewMusicPlayer(musicCodec.SampleRate(), musicCodec.Channels(), music.Browsers())
	c.context = &agent.Context{
		Announce:    c.Announce,
		MusicPlayer: player,
		MusicGroup:  c.musicGroup,
	}
	return nil
}

// -- Transport callbacks --

func (c *DeviceConnection) onEvent(event device.EventType, payload []byte) {
	switch event {
	case device.EventWakeWord:
		c.handleWake(payload)
	case device.EventMicConfig:
		c.handleMicConfig(payload)
	}
}

func (c *DeviceConnection) handleWake(payload []byte) {
	c.cancelTask()

	// Pause group music on ALL devices when any device wakes
	if c.musicGroup != nil && c.musicGroup.IsActive() {
		c.musicGroup.Pause(context.Background())
	}

	c.mu.Lock()
	continuing := c.continuing
	c.continuing = false
	c.audioQueue = make(chan []byte, micQueueSize)
	c.mu.Unlock()

	if continuing {
		// Follow-up — already committed to this device, skip arbitration.
		slog.Info("Follow-up session started")
		c.startTask(c.run)
		return
	}

	// Parse extended payload: [2B LE peak][2B LE ambient][wake_word UTF-8]
	// The normalised score (peak/ambient) cancels out mic gain differences
	// so the *closest* device wins, not the loudest mic.
	score := 0
	wakeWord := ""
	if len(payload) >= 4 {
		peak := int(binary.LittleEndian.Uint16(payload[0:]))
		ambient := int(binary.LittleEndian.Uint16(payload[2:]))
		wakeWord = string(payload[4:])
		if ambient < 1 {
			ambient = 1
		}
		score = peak * 1000 / ambient
		slog.Debug("Wake payload", "peak", peak, "ambient", ambient, "score", score)
	} else if len(payload) > 0 {
		wakeWord = string(payload)
	}

	if c.onWake != nil {
		c.onWake(c, score, wakeWord)
		return
	}
	// No arbitration — start immediately.
	c.pipeline.ResetHistory()
	slog.Info("Voice session started", "wake_word", wakeWord)
	c.startTask(c.run)
}

func (c *DeviceConnection) handleMicConfig(payload []byte) {
	if len(payload) < 7 {
		return
	}
	rate := int(binary.LittleEndian.Uint32(payload[0:]))
	nbyte := int(binary.LittleEndian.Uint16(payload[4:]))
	name := codec.NameForID(payload[6])

	// Use the device's reported mic codec
	mic, err := codec.NewCodec(string(name), rate, 1, nbyte)
	if err != nil {
		slog.Error("Cannot create mic codec", "codec", name, "err", err)
		return
	}
	slog.Info("Mic audio config (device)", "device", c.Name, "codec", codec.Describe(mic))

	// If the server prefers a different mic codec, request it
	preferred := c.settings.Transport.Codec
	if preferred != string(name) {
		pref, err := codec.NewCodec(preferred, rate, 1, 0)
		if err == nil {
			buf := make([]byte, 7)
			binary.LittleEndian.PutUint32(buf[0:], uint32(rate))
			binary.LittleEndian.PutUint16(buf[4:], uint16(pref.EncodedFrameBytes()))
			buf[6] = pref.ID()
			if err := c.transport.SendEvent(device.EventMicConfig, buf); err == nil {
				mic = pref
				slog.Info("Mic audio config (requested)", "device", c.Name, "codec", codec.Describe(pref))
			}
		}
	}
	c.mu.Lock()
	c.micCodec = mic
	c.mu.Unlock()
}

// onAudio decodes mic audio and queues PCM for the pipeline.
func (c *DeviceConnection) onAudio(data []byte) {
	c.mu.Lock()
	mic := c.micCodec
	queue := c.audioQueue
	c.mu.Unlock()
	pcm := data
	if mic != nil {
		decoded, err := mic.Decode(data)
		if err != nil {
			slog.Debug("Mic decode failed", "err", err)
			return
		}
		pcm = decoded
	}
	select {
	case queue <- pcm:
	default:
		slog.Debug("Mic queue full, dropping frame", "device", c.Name)
	}
}

func (c *DeviceConnection) onDisconnect() {
	c.mu.Lock()
	cancel := c.taskCancel
	c.mu.Unlock()
	if cancel != nil {
		cancel()
	}
}

func (c *DeviceConnection) onReconnect() {
	time.Sleep(c.SetupDelay)
	if err := c.setupDevice(); err != nil {
		slog.Error("Device setup after reconnect failed", "device", c.Name, "err", err)
	}
}

// -- Pipeline execution --

// run executes one voice session on this device.
func (c *DeviceConnection) run(ctx context.Context) {
	c.mu.Lock()
	mic := c.audioQueue
	actx := c.context
	c.mu.Unlock()

	c.output.Reset()
	continuing := c.pipeline.Run(ctx, c.output, mic, actx)
	_ = c.output.Flush(ctx)

	c.mu.Lock()
	c.continuing = continuing
	c.mu.Unlock()

	// After the voice pipeline completes, resume music if it was playing
	if !continuing && actx != nil && ctx.Err() == nil {
		if c.musicGroup != nil && c.musicGroup.IsActive() {
			// Group music was paused for voice — resume on all devices
			c.musicGroup.Resume(ctx)
		} else if actx.MusicPlayer != nil && actx.MusicPlayer.IsActive() {
			c.runMusic(ctx, actx.MusicPlayer)
		}
	}
}

// runMusic streams music at 48 kHz via the dedicated music output.
func (c *DeviceConnection) runMusic(ctx context.Context, player *music.MusicPlayer) {
	out := c.musicOutput
	out.Reset()
	if err := out.SendEvent(ctx, device.EventTTSStart, nil); err != nil {
		return
	}
	err := player.Stream(ctx, out)
	// Always try to close the speaker, even after cancellation.
	tail, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if ferr := out.Flush(tail); ferr == nil {
		_ = out.SendEvent(tail, device.EventTTSEnd, nil)
	}
	if err != nil && ctx.Err() == nil {
		slog.Error("Music streaming error", "device", c.Name, "err", err)
	}
}

func (c *DeviceConnection) runAnnounce(text string) func(ctx context.Context) {
	return func(ctx context.Context) {
		c.output.Reset()
		c.pipeline.Announce(ctx, c.output, text)
		_ = c.output.Flush(ctx)
	}
}

// -- Helpers --

// startTask cancels any running task and starts fn in a new goroutine.
func (c *DeviceConnection) startTask(fn func(ctx context.Context)) {
	c.cancelTask()
	ctx, cancel := context.WithCancel(context.Background())
	done := make(chan struct{})
	c.mu.Lock()
	c.taskCancel = cancel
	c.taskDone = done
	c.mu.Unlock()
	go func() {
		defer close(done)
		fn(ctx)
	}()
}

// cancelTask cancels the running task and waits for it to finish.
func (c *DeviceConnection) cancelTask() {
	c.mu.Lock()
	cancel := c.taskCancel
	done := c.taskDone
	c.taskCancel = nil
	c.taskDone = nil
	c.mu.Unlock()
	if cancel != nil {
		cancel()
		// The task may be blocked on the paced output; reset it so it
		// unwinds promptly, then wait.
		c.output.Reset()
		<-done
	}
}

// TaskRunning reports whether a session/announcement is in progress.
func (c *DeviceConnection) TaskRunning() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.taskDone == nil {
		return false
	}
	select {
	case <-c.taskDone:
		return false
	default:
		return true
	}
}

// StartPipeline is called by DeviceManager when this device wins arbitration.
func (c *DeviceConnection) StartPipeline(wakeWord string) {
	c.pipeline.ResetHistory()
	slog.Info("Voice session started (won arbitration)", "device", c.Name, "wake_word", wakeWord)
	c.startTask(c.run)
}

// AbortWake is called by DeviceManager when this device loses arbitration.
func (c *DeviceConnection) AbortWake() {
	c.cancelTask()
	// Drain any buffered mic audio
	c.mu.Lock()
	queue := c.audioQueue
	c.mu.Unlock()
	for {
		select {
		case <-queue:
			continue
		default:
		}
		break
	}
	_ = c.transport.SendEvent(device.EventWakeAbort, nil)
	slog.Info("Wake aborted (another device won)", "device", c.Name)
}

// Announce streams a TTS announcement directly to the device, interrupting
// any running session.
func (c *DeviceConnection) Announce(text string) {
	c.startTask(c.runAnnounce(text))
}

// AnnounceAndWait announces and blocks until playback finishes or ctx ends.
func (c *DeviceConnection) AnnounceAndWait(ctx context.Context, text string) error {
	c.Announce(text)
	c.mu.Lock()
	done := c.taskDone
	c.mu.Unlock()
	if done == nil {
		return nil
	}
	select {
	case <-done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
